app/src/androidTest/java/org/mozilla/tv/firefox/ui/YouTubeNavigationTest.kt

```diff
@@ -0,0 +1,134 @@
+    private lateinit var server: MockWebServer
+    private lateinit var pages: List<TestAssetHelper.TestAsset>
+
+    /* ktlint-disable no-blank-line-before-rbrace */ // This imposes unreadable grouping.
+
+    @Before
+    fun setup() {
+        server = MockWebServer().apply {
+            setDispatcher(AndroidAssetDispatcher())
+            start()
+        }
+
+        pages = TestAssetHelper.getGenericAssets(server).dropLast(1) +
+            TestAssetHelper.getUUIDPage(server)
+    }
+
+    /**
+     * Test page 1, YouTube TV, Test page 2, back to YouTube, back out of Youtube
+     * Expected: Test page 1
+     */
+    @Test
+    fun youtubeBackTest1() {
+        navigationOverlay {
+        }.enterUrlAndEnterToBrowser(pages[0].url) {
+        }.openOverlay {
+        }.enterUrlAndEnterToBrowser("youtube.com/tv".toUri()!!) {
+        }.openOverlay {
+        }.enterUrlAndEnterToBrowser(pages[1].url) {
+            remoteBack()
+        }.openOverlay {
+            assertURLBarTextContains("youtube")
+        }.closeToBrowser {
+            backOutOfYouTube()
+        }.openOverlay {
+            assertURLBarTextContains(pages[0].url.toString())
+        }
+    }
+    /**
+     * YouTube tile, Test page 1, back to YouTube, Test page 2, back to YouTube, back out of YouTube
+     * Expected: Overlay
+     */
+    @Test
+    fun youtubeBackTest2() {
+        navigationOverlay {
+        }.openTileToBrowser(0) {
+        }.openOverlay {
+        }.enterUrlAndEnterToBrowser(pages[0].url) {
+            remoteBack()
+        }.openOverlay {
+        }.enterUrlAndEnterToBrowser(pages[1].url) {
+            remoteBack()
+        }.openOverlay {
+            assertURLBarTextContains("youtube")
+        }.closeToBrowser {
+            backOutOfYouTube()
+        }.openOverlay {
+            assertURLBarDisplaysHint()
+        }
+    }
+    /**
+     * YouTube from Pocket, back out of YouTube, YouTube TV, back out of YouTube
+     * Expected: Overlay
+     */
+    @Test
+    fun youtubeBackTest3() {
+        navigationOverlay {
+        }.openPocketMegatile {
+        }.openTileToBrowser(0) {
+            remoteBack()
+            // Skip the sign-in prompt
+            onWebView().withElement(findElement(Locator.CLASS_NAME, "overlay-partials-menu-item"))
+                .perform(webClick())
+            backOutOfYouTube()
+        }
+        navigationOverlay {
+            assertURLBarDisplaysHint()
+        }.enterUrlAndEnterToBrowser("youtube.com/tv".toUri()!!) {
+            backOutOfYouTube()
+        }.openOverlay {
+            assertURLBarDisplaysHint()
+        }
+    }
+
+    /**
+     * Clicks remote back until the active element is in the YouTube sidebar.
+     * Then, clicks remote back one more time to go to the previous site.
+     */
+    private fun backOutOfYouTube() {
+        browser {
+            var activeElementParentId = executeJS(
+                "return document.activeElement.parentElement.parentElement.id"
```

Comment:
These tests pass on device, but are failing on emulator at this line with:
`Cannot read property 'id' of null`

@Baron-Severin Do you of any differences between device and emulator that could be causing this?

Reply:
I've run this branch a few times now on both device and emulator and haven't encountered that failure, so there seems to be some flakiness here.  

If the grandparent of the focused element is null, it means the focused element is `body` (the root element is `html`, which is one level higher than `body`).  You can verify that by logging the active element.  I know that `body` is the element focused by default until YouTube has finished loading, after which I assume they programmatically request focus on the first video.  My guess is that our test is sometimes trying to move on before YouTube is fully loaded.

Reply:
So do you think the best thing would be to 
a) change the test to check `noElementFocused || bodyElementFocused || sidebarFocused` like in the implementation
b) change the test to make sure it waits for YT to be loaded (not entirely sure how to accomplish this since it's already using `SessionLoadedIdlingResource` which should be waiting)

Reply:
B, although I'm not sure how to accomplish it.  The `SessionLoadedIdlingResource` knows whether or not the page load is complete, but maybe the focus change itself happens on page load complete?  So it becomes a race condition between the two?  I'm not sure, it'll take some investigation.

I think one option here would be to break the Espresso test into a separate PR so we can avoid blocking the other story.